### PR TITLE
Rosetta fixes

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
0xs were inconsistent across types. Coins didn't have the associated move type, so it made it confusing.  Now its' added.

Also, gas wasn't always charged so it caused issues during failed transactions.  Now, all gas is always charged all the time.

### Test Plan
```
$ rosetta-cli check:data --configuration-file rosetta_config.json

Success: Reconciliation Coverage End Condition [Coverage: 100.000000%]

+--------------------+--------------------------------+--------+
|  CHECK:DATA TESTS  |          DESCRIPTION           | STATUS |
+--------------------+--------------------------------+--------+
| Request/Response   | Rosetta implementation         | PASSED |
|                    | serviced all requests          |        |
+--------------------+--------------------------------+--------+
| Response Assertion | All responses are correctly    | PASSED |
|                    | formatted                      |        |
+--------------------+--------------------------------+--------+
| Block Syncing      | Blocks are connected into a    | PASSED |
|                    | single canonical chain         |        |
+--------------------+--------------------------------+--------+
| Balance Tracking   | Account balances did not go    | PASSED |
|                    | negative                       |        |
+--------------------+--------------------------------+--------+
| Reconciliation     | No balance discrepancies were  | PASSED |
|                    | found between computed and     |        |
|                    | live balances                  |        |
+--------------------+--------------------------------+--------+

+--------------------------+--------------------------------+-------------+
|     CHECK:DATA STATS     |          DESCRIPTION           |    VALUE    |
+--------------------------+--------------------------------+-------------+
| Blocks                   | # of blocks synced             |       14482 |
+--------------------------+--------------------------------+-------------+
| Orphans                  | # of blocks orphaned           |           0 |
+--------------------------+--------------------------------+-------------+
| Transactions             | # of transaction processed     |       28825 |
+--------------------------+--------------------------------+-------------+
| Operations               | # of operations processed      |          44 |
+--------------------------+--------------------------------+-------------+
| Accounts                 | # of accounts seen             |           6 |
+--------------------------+--------------------------------+-------------+
| Active Reconciliations   | # of reconciliations performed |          31 |
|                          | after seeing an account in a   |             |
|                          | block                          |             |
+--------------------------+--------------------------------+-------------+
| Inactive Reconciliations | # of reconciliations performed |         314 |
|                          | on randomly selected accounts  |             |
+--------------------------+--------------------------------+-------------+
| Exempt Reconciliations   | # of reconciliation failures   |           0 |
|                          | considered exempt              |             |
+--------------------------+--------------------------------+-------------+
| Failed Reconciliations   | # of reconciliation failures   |           0 |
+--------------------------+--------------------------------+-------------+
| Skipped Reconciliations  | # of reconciliations skipped   |           0 |
+--------------------------+--------------------------------+-------------+
| Reconciliation Coverage  | % of accounts that have been   | 100.000000% |
|                          | reconciled                     |             |
+--------------------------+--------------------------------+-------------+

badger 2022/07/11 16:15:36 INFO: Storing value log head: {Fid:0 Len:33 Offset:19906543}
```

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/1892)

### CODE REVIEW

File: api/src/context.rs [hunk 1]
```diff
@@ -131,14 +130,13 @@ impl Context {
...
 
     /// Retrieves information about a block
-    pub fn get_block_info(&self, version: u64, ledger_version: u64) -> Result<Option<BlockInfo>> {
+    pub fn get_block_info(&self, version: u64, ledger_version: u64) -> Result<BlockInfo> {
         // We scan the DB to get the block boundaries
         let (start, end) = match self.db.get_block_boundaries(version, ledger_version) {
             Ok(inner) => inner,
             Err(error) => {
-                debug!("Failed to get block boundaries {}", error);
                 // None means we can't find the block
```
Comment (banool) [inline api/src/context.rs:138]: nit: Comment out of date.

File: crates/aptos-rosetta/src/account.rs [hunk 2]
```diff
@@ -179,7 +182,14 @@ async fn get_balances(
             })
             .collect())
     } else {
-        Ok(HashMap::new())
+        let mut currency_map = HashMap::new();
+        currency_map.insert(
+            native_coin_tag(),
```
Comment (banool) [inline crates/aptos-rosetta/src/account.rs:187]: This makes me wonder if we should just do this in both arms. As in, do something like:
```
currency_map.entry(native_coin_tag()).or_insert_with(|| Balance {coin: TestCoin {value: U64(0) }});
```
Just before the return.

Might be weird to assume we have `native_coin_tag` by default only in some cases. Correct me if the previous block is doing something I'm missing.

File: crates/aptos-rosetta/src/block.rs [hunk 3]
```diff
@@ -151,14 +151,12 @@ impl BlockCache {
         let mut versions = BTreeMap::new();
         // Genesis is always index 0
         // TODO: Ensure that this won't fail if it's been pruned
```
Comment (banool) [inline crates/aptos-rosetta/src/block.rs:153]: Where are we at with this, does the genesis get pruned? Seems like maybe we'd want to keep it around always.

File: crates/aptos-rosetta/src/network.rs [hunk 4]
```diff
@@ -148,39 +147,23 @@ async fn network_status(
...
+    // Get the oldest block
+    let oldest_block_identifier = if let Some(version) = state.oldest_ledger_version {
+        let block_info = block_cache.get_block_info_by_version(version).await?;
+        Some(BlockIdentifier::from_block_info(block_info))
     } else {
-        return Err(ApiError::BlockIncomplete);
+        None
     };
 
+    // Get the latest block
```
Comment (banool) [inline crates/aptos-rosetta/src/network.rs:158]: nit: Incorrect comment?

File: crates/aptos-rosetta/src/types/identifiers.rs [hunk 5]
```diff
@@ -207,7 +207,7 @@ pub struct TransactionIdentifier {
 impl From<&TransactionInfo> for TransactionIdentifier {
     fn from(txn: &TransactionInfo) -> Self {
         TransactionIdentifier {
-            hash: txn.accumulator_root_hash.to_string(),
+            hash: strip_hex_prefix(&txn.hash.to_string()).to_string(),
```
Comment (banool) [inline crates/aptos-rosetta/src/types/identifiers.rs:210]: Not necessary in this PR but I wonder if the hash types should have `with_prefix` and `without_prefix` functions to make this a little easier.

**Review by banool — APPROVED** [review]
Okay given what we talked about today with how block boundaries are defined this looks legit.

File: crates/aptos-rosetta/src/block.rs (hunk 3 above)
Comment (gregnazario) [reply crates/aptos-rosetta/src/block.rs:153]: Yes, I need to look at a way to get the hash for this (when it's pruned) and as the comment above implies, this needs to be checked.h

**Pull request merged**